File: web/src/lib/bedrock.ts
import {
  BedrockRuntimeClient,
  ConverseCommand,
  type Message as BedrockMessage,
} from '@aws-sdk/client-bedrock-runtime';

let _client: BedrockRuntimeClient | null = null;

function getClient(): BedrockRuntimeClient {
  if (!_client) {
    _client = new BedrockRuntimeClient({
      region: process.env.AWS_REGION ?? 'us-east-1',
      credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
        ...(process.env.AWS_SESSION_TOKEN && {
          sessionToken: process.env.AWS_SESSION_TOKEN,
        }),
      },
    });
  }
  return _client;
}

const MODEL_ID: string =
  process.env.BEDROCK_MODEL_ID ?? 'us.anthropic.claude-sonnet-4-6-20250514-v1:0';

export async function converse(
  systemPrompt: string,
  messages: BedrockMessage[],
): Promise<string> {
  const command = new ConverseCommand({
    modelId: MODEL_ID,
    system: [{ text: systemPrompt }],
    messages,
    inferenceConfig: {
      maxTokens: 512,
      temperature: 0.7,
    },
  });

  const response = await getClient().send(command);
  const content = response.output?.message?.content?.[0];
  if (content && 'text' in content && content.text) return content.text;
  throw new Error('No text output from Bedrock');
}

File: backend/src/lib/bedrock.py
import os
import boto3
from typing import List, Optional

_client = None

MODEL_ID: str = os.environ.get(
    "BEDROCK_MODEL_ID", "us.anthropic.claude-sonnet-4-6-20250514-v1:0"
)


def get_client():
    global _client
    if _client is None:
        kwargs: dict = {
            "region_name": os.environ.get("AWS_REGION", "us-east-1"),
            "aws_access_key_id": os.environ.get("AWS_ACCESS_KEY_ID"),
            "aws_secret_access_key": os.environ.get("AWS_SECRET_ACCESS_KEY"),
        }
        session_token = os.environ.get("AWS_SESSION_TOKEN")
        if session_token:
            kwargs["aws_session_token"] = session_token
        _client = boto3.client("bedrock-runtime", **kwargs)
    return _client


def converse(system_prompt: str, messages: List[dict], max_tokens: int = 512) -> str:
    response = get_client().converse(
        modelId=MODEL_ID,
        system=[{"text": system_prompt}],
        messages=messages,
        inferenceConfig={"maxTokens": max_tokens, "temperature": 0.7},
    )
    content_blocks = response.get("output", {}).get("message", {}).get("content", [])
    if content_blocks and "text" in content_blocks[0]:
        return content_blocks[0]["text"]
    raise ValueError("No text output from Bedrock")
